src/app/public/src/modules/clipboard/clipboard.service.ts

```diff
@@ -0,0 +1,89 @@
+// adapted from ngx-clipboard
+
+import { Inject, Injectable, Renderer2 } from '@angular/core';
+import { DOCUMENT } from '@angular/common';
+import { StacheWindowRef } from '../../..';
+
+@Injectable()
+export class StacheClipboardService {
+  private tempTextArea: HTMLTextAreaElement | undefined;
+
+  constructor(
+    @Inject(DOCUMENT) private document: any,
+    private windowRef: StacheWindowRef
+  ) { }
+
+  /**
+   * copyFromInputElement
+   */
+  public copyFromInputElement(targetElm: HTMLInputElement | HTMLTextAreaElement, renderer: Renderer2): boolean {
+    try {
```

Comment:
Avoid `try/catch` and only check known conditions. The reason is that `try/catch` could potentially handle an error you didn't intend for--such as an invalid type--(and your unit test should have caught to begin with). Instead, try to determine what the potential error could be, and let unknown errors bubble out naturally.

See: https://www.dotnetperls.com/tester-doer